homeassistant/components/climate/myicomfort.py

```diff
@@ -0,0 +1,225 @@
+    username: YOUR_MYICOMFORT_USERNAME
+    password: YOUR_MYICOMFORY_PASSWORD
+
+"""
+
+import logging
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+
+from homeassistant.components.climate import (
+    ATTR_TARGET_TEMP_HIGH, ATTR_TARGET_TEMP_LOW,
+    ClimateDevice, PLATFORM_SCHEMA, STATE_AUTO,
+    STATE_COOL, STATE_HEAT, SUPPORT_TARGET_TEMPERATURE,
+    SUPPORT_TARGET_TEMPERATURE_HIGH, SUPPORT_TARGET_TEMPERATURE_LOW,
+    SUPPORT_OPERATION_MODE, SUPPORT_AWAY_MODE, SUPPORT_FAN_MODE)
+from homeassistant.const import (
+    CONF_USERNAME, CONF_PASSWORD, TEMP_CELSIUS, TEMP_FAHRENHEIT,
+    STATE_ON, STATE_OFF, STATE_IDLE, ATTR_TEMPERATURE)
+
+REQUIREMENTS = ['myicomfort==0.2.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+# HA doesn't have a 'circulate' state defined.
+STATE_CIRCULATE = 'circulate'
+
+SUPPORT_FLAGS = (SUPPORT_TARGET_TEMPERATURE |
+                 SUPPORT_TARGET_TEMPERATURE_HIGH |
+                 SUPPORT_TARGET_TEMPERATURE_LOW |
+                 SUPPORT_OPERATION_MODE |
+                 SUPPORT_AWAY_MODE |
+                 SUPPORT_FAN_MODE)
+
+FAN_MODES = [
+    STATE_AUTO, STATE_ON, STATE_CIRCULATE
+]
+
+OP_MODES = [
+    STATE_OFF, STATE_HEAT, STATE_COOL, STATE_AUTO
+]
+
+SYSTEM_STATES = [
+    STATE_IDLE, STATE_HEAT, STATE_COOL
+]
+
+TEMP_UNITS = [
+    TEMP_FAHRENHEIT, TEMP_CELSIUS
+]
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional('name', default='iComfort'): cv.string,
+    vol.Optional('system', default=0): vol.Coerce(int),
+    vol.Optional('zone', default=0): vol.Coerce(int),
+    vol.Optional('min_temp'): vol.Coerce(float),
+    vol.Optional('max_temp'): vol.Coerce(float),
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the climate platform."""
+    from myicomfort.api import Tstat
+
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    system = config.get('system')
+    zone = config.get('zone')
+    name = config.get('name')
+    min_temp = config.get('min_temp')
+    max_temp = config.get('max_temp')
+
+    climate = [MyiComfortClimate(name, min_temp, max_temp,
+                                 Tstat(username, password, system, zone))]
+
+    add_devices(climate, True)
+
+
+class MyiComfortClimate(ClimateDevice):
+    """Class for Lennox iComfort WiFi thermostat."""
+
+    def __init__(self, name, min_temp, max_temp, api):
+        """Initialize the climate device."""
+        self._name = name
+        self._api = api
+        self._min_temp = min_temp
+        self._max_temp = max_temp
+
+    def update(self):
+        """Update data from the thermostat API."""
+        self._api.pull_status()
+
+    @property
+    def device_state_attributes(self):
+        """Return device specific state attributes."""
+        return {
+        }
+
+    @property
+    def state(self):
```

Comment:
Don't overwrite `state`. It is the responsibility of the base climate entity. Use `current_operation`.

Reply:
I am confused on this item.  My understanding is that `state` is the current running state of the HVAC system (idle, heating, cooling) and `current_operation` is the operation mode selected on the thermostat (off, heat, cool, heat&cool). Did I miss a property that would inform the base climate entity what the current running state is?

Reply:
Read the code for the base and did some digging thru my notes. I guess I assumed that it was just a place holder and something I should override if I wanted mode details.  I based the my code on someone else's code and they based there's on the generic thermostat which overrides state. 

To me it seemed very beneficial to override as my thermostat is in Auto mode for half the year (Thank you Texas weather).  During that time I have HA do things based on whether the system has been cooling vs heating that day. With the behavior of the base entity that data is lost. It does make more sense now why the Lovelace thermostat card handles state the way it does.

So my question becomes, is it a requirement to abide by this rule or can we override state when it provides more data to the system?

Reply:
It is required to follow our architecture. See the architecture repo issue 22 for the plans for the climate component.

Reply:
I read thru that thread and I understand they are rethinking the climate design to address the split between thermostat state and actual device state. I can see the architecture repo has a lot of good info to absorb.  Is there a place that gives a breakdown of component requirements, options, and rules so I don't try overriding the wrong thing in the future?

I will drop the state override as required and I guess I could throw the HVAC status into that empty device_state_attributes property.

Reply:
Read our developer docs. But they are not complete in terms of rules and standards unfortunately. We welcome improvements in this regard.

https://developers.home-assistant.io/